modules/distributor/queue/queue.go

```diff
@@ -0,0 +1,224 @@
+package queue
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/weaveworks/common/user"
+	"go.uber.org/atomic"
+)
+
+type request[T any] struct {
+	data T
+}
+
+type ProcessFunc[T any] func(ctx context.Context, data T) error
+
+// Queue represents a single tenant's queue.
+type Queue[T any] struct {
+	// wg is used to wait for the workers to drain the queue while stopping
+	wg sync.WaitGroup
+
+	logger log.Logger
+
+	name           string
+	tenantID       string
+	workerCount    int
+	size           int
+	reqChan        chan *request[T]
+	fn             ProcessFunc[T]
+	workersCloseCh chan struct{}
+
+	pushesTotalMetrics         *prometheus.CounterVec
+	pushesFailuresTotalMetrics *prometheus.CounterVec
+	lengthMetric               *prometheus.GaugeVec
+
+	readOnly *atomic.Bool
+}
+
+func New[T any](cfg Config, logger log.Logger, reg prometheus.Registerer, fn ProcessFunc[T]) *Queue[T] {
+	pushesTotalMetrics := prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "tempo",
+		Subsystem: "distributor",
+		Name:      "queue_pushes_total",
+		Help:      "Total number of successful requests queued up for a tenant to the generatorForwarder",
+	}, []string{"name", "tenant"})
+	pushesFailuresTotalMetric := prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "tempo",
+		Subsystem: "distributor",
+		Name:      "queue_pushes_failures_total",
+		Help:      "Total number of failed pushes to the queue for a tenant to the generatorForwarder",
+	}, []string{"name", "tenant"})
+	lengthMetric := prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "tempo",
+		Subsystem: "distributor",
+		Name:      "queue_length",
+		Help:      "Number of queued requests for a tenant",
+	}, []string{"name", "tenant"})
+
+	var alreadyRegisteredErr prometheus.AlreadyRegisteredError
+	err := reg.Register(pushesTotalMetrics)
+	if err != nil && !errors.As(err, &alreadyRegisteredErr) {
```

Comment:
does this work? if we fail to register this metric a second time due to it being already registered then we lose a reference to the original metric that is registered and these metrics stop updating? this will happen the first time the config is reloaded?

just define them with promauto?

Reply:
I think the only easy way to fix this is to define those metrics on the package level to make initialize them only once. It's a bummer, but I see no other way around this without an abstraction around Prometheus. Therefore, I moved them outside of this function and they are now "global".